### PR TITLE
GeoRedundantReadMode

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Addresses #21445

Adds GeoRedundantReadMode enum to Azure.Storage.Common and adds a field for it to relevant packages' ClientOptions types.

Adds a check to Azure.Core.Argument (shared source) for if given enum value is defined in the type.

### CODE REVIEW

**Comment by jaschrep-msft** [comment]
"Do Not Merge" added as this PR contains API changes and the upcoming release has already gone through API review. Must wait for next one.

**Comment by jaschrep-msft** [comment]
TODO: changelog after version bump from upcoming release

File: sdk/core/Azure.Core/src/Shared/Argument.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/Argument.cs:194]:
```suggestion
                throw new ArgumentException($"Value not defined in enum {enumType.FullName}.", name);
```

File: sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs:24]: What would be the point of setting PrimaryOnly? Isn't this equivalent to leaving the secondary Uri null?

Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs:24]: Really just a parity inclusion. Not needed. Technically, secondary only isn't needed either, as the clients could just be instantiated with secondary from the start. We could get away with just the first two enum values.

They were both included as the original feature request was to port the enum and its functionality from v11, which had these values. But also v11 had a full `StorageUri` class that contained primary and secondary together, so it made a bit more sense back then.

**Review by seanmcc-msft — APPROVED** [review]

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs:24]: Discussed separately with team. Will be removing PrimaryOnly and SecondaryOnly modes, as they are possible by just creating the client with the desired host. Adding support later, if necessary, would not be a breaking change.

File: sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs:9]: Would it make any sense to include `None` as a default value?

**Review by seanmcc-msft — COMMENTED** [review]
Does secondary endpoint only apply for the blob, DFS, and queue endpoints?

File: sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/GeoRedundantReadMode.cs:9]: The SDK is already in PrimaryThenSecondary mode today, and just doesn't bother if the secondary endpoint wasn't provided. I don't know what None would help convey here.

**Comment by jaschrep-msft** [comment]
> Does secondary endpoint only apply for the blob, DFS, and queue endpoints?

Yes. File Shares do not have secondary endpoints.

**Review by seanmcc-msft — APPROVED** [review]

**Comment by jaschrep-msft** [comment]
@JoshLove-msft or @jsquire can I get an approval from one of you regarding the azure-core helper method being added?

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**
